### PR TITLE
[aap_controller] Implementing AAP Automation Controller Plugin

### PR DESCRIPTION
**Pull request by lucas-benedito**
Adding the file 'aap_controller.py' for the sos report collects
The files used for troubleshooting issues at
Ansible Automation Platform Automation Controller

Related: RH AAP-18409

Signed-off-by: Lucas Benedito <lbenedit@redhat.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

### CODE REVIEW

File: sos/report/plugins/aap_controller.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Copyright (c) 2023 Pavel Moravec <pmoravec@redhat.com>
```
Comment (TurboTurtle) [inline sos/report/plugins/aap_controller.py:1]: Copy error?

File: sos/report/plugins/aap_controller.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    short_desc = 'AAP Automation Controller plugin'
+    plugin_name = 'aap_controller'
+    profiles = ('sysmgmt', 'ansible')
+
+    packages = ('automation-controller-venv-tower',
+                'automation-controller-server',
+                'automation-controller-ui',
+                'automation-controller')
+    commands = ('awx-manage')
+    services = ('automation-controller')
```
Comment (TurboTurtle) [inline sos/report/plugins/aap_controller.py:25]: These need to be tuples.

File: sos/report/plugins/aap_controller.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            "/var/log/supervisor",
+            "/var/log/unattended-upgrades",
+        ])
+
+        self.add_forbidden_path([
+            "/etc/tower/SECRET_KEY",
+            "/etc/tower/tower.key",
+            "/etc/tower/awx.key",
+            "/etc/tower/tower.cert",
+            "/etc/tower/awx.cert",
```
Comment (TurboTurtle) [inline sos/report/plugins/aap_controller.py:45]: Should these not be expanded to something like `/etc/tower/*.key` and `/etc/tower/*.cert`? I don't know if it's possible for other filenames to be used here, but if so I imagine we don't want to copy any of them.

File: sos/report/plugins/aap_controller.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            "/var/lib/awx/venv/awx/bin/pip freeze -l"
+            "/var/lib/awx/venv/ansible/bin/pip freeze"
+            "/var/lib/awx/venv/ansible/bin/pip freeze -l"
+            "tree -d /var/lib/awx"
+            "ls -ll /var/lib/awx"
+            "ls -ll /var/lib/awx/venv"
+            "ls -ll /etc/tower",
+            "ls -ll /var/run/awx-receptor"
+            "ls -ll /etc/receptor",
+            "receptorctl --socket /var/run/awx-receptor/receptor.sock status"
```
Comment (TurboTurtle) [inline sos/report/plugins/aap_controller.py:70]: Aren't these collected by the plugin proposed in #3553?

File: sos/report/plugins/aap_controller.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+        self.add_forbidden_path([
+            "/etc/tower/SECRET_KEY",
+            "/etc/tower/tower.key",
+            "/etc/tower/awx.key",
+            "/etc/tower/tower.cert",
+            "/etc/tower/awx.cert",
+            "/var/log/tower/profile",
+            "/etc/receptor/tls/ca/*.key",
+            "/etc/receptor/tls/*.key",
```
Comment (TurboTurtle) [inline sos/report/plugins/aap_controller.py:48]: Aren't these also collected by the proposed receptor plugin in #3553?

**Comment by jcastill** [comment]
Thanks for the new plugin @lucas-benedito ! A couple of things:
- Could you squash all four commits into one?
- Could you sign the resulting commit?

File: sos/report/plugins/aap_controller.py (hunk 1 above)
Comment (pmoravec) [reply sos/report/plugins/aap_controller.py:1]: I can leave with it.. :)

In fact, the plugin was internally prepared in a collaborative manner by few fellows, also with my contribution. I would vote for putting here my and Lucas's name, but I am ok with either option.

File: sos/report/plugins/aap_controller.py (hunk 3 above)
Comment (pmoravec) [reply sos/report/plugins/aap_controller.py:45]: I *think* - in general - that all keys (private part of cert) should be forbidden, while there can be some reason to copy a cert file (public part, yet internally for the deployment) in some cases. Don't know how this applies to the specific keys / certs in `/etc/tower`.

File: sos/report/plugins/aap_controller.py (hunk 1 above)
Comment (lucas-benedito) [reply sos/report/plugins/aap_controller.py:1]: @pmoravec++ made the initial draft and i just adjusted some points

File: sos/report/plugins/aap_controller.py (hunk 5 above)
Comment (lucas-benedito) [reply sos/report/plugins/aap_controller.py:48]: Ack, will remove the receptor references. thanks

File: sos/report/plugins/aap_controller.py (hunk 4 above)
Comment (lucas-benedito) [reply sos/report/plugins/aap_controller.py:70]: Ack, will remove the receptor references. thanks

File: sos/report/plugins/aap_controller.py (hunk 3 above)
Comment (lucas-benedito) [reply sos/report/plugins/aap_controller.py:45]: Changed it as suggested.

File: sos/report/plugins/aap_controller.py (hunk 2 above)
Comment (lucas-benedito) [reply sos/report/plugins/aap_controller.py]: done

File: sos/report/plugins/aap_controller.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    services = ('automation-controller',)
+
+    def setup(self):
+        self.add_copy_spec([
+            "/etc/tower/",
+            "/etc/supervisord.conf",
+            "/etc/supervisord.d/*",
+            "/var/log/tower",
+            "/var/log/nginx/automationcontroller.access.log*",
+            "/var/log/nginx/automationcontroller.error.log*",
```
Comment (pmoravec) [inline sos/report/plugins/aap_controller.py:34]: These file patterns should be forbidden in `nginx` plugin (did I write a comment about it here or to a different PR? I cant find it now..)

**Review by pmoravec — APPROVED** [review]
LGTM, thanks!

Just please amend the commit to follow DCO (signed-off and ideally also Closes: #3561).

**Review by TurboTurtle — APPROVED** [review]

File: sos/report/plugins/aap_controller.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            "/var/log/tower/profile",
+        ])
+
+        self.add_cmd_output([
+            "awx-manage --version"
+            "awx-manage list_instances"
+            "awx-manage run_dispatcher --status"
+            "awx-manage run_callback_receiver --status"
+            "awx-manage check_license --data"
+            "awx-manage run_wsbroadcast --status"
```
Comment (TurboTurtle) [inline sos/report/plugins/aap_controller.py:52]: Missing commas throughout the list

**Review by TurboTurtle — CHANGES_REQUESTED** [review]
Comments above, from codeQL scan. Good sanity check here!

**Comment by jskacel** [comment]
@lucas-benedito can we add also `podman images` and `podman ps`? Would be helpful I believe to see that. Maybe `podman info` as well.

**Comment by arif-ali** [comment]
Curious on this, and I could be wrong here.

Isn't Ansible Tower/AWX something that can be used anywhere else. It seems like these are the main things we are collecting here. i.e. AWX specifically could be potentially be used on other distros, so we could make the name of the plugin more generic so that it can be used in other distros if needed in the future. I know ubuntu doesn't have packages yet, but could happen

**Comment by TurboTurtle** [comment]
> @lucas-benedito can we add also `podman images` and `podman ps`? Would be helpful I believe to see that. Maybe `podman info` as well.

These are already captured by the `podman` plugin.

**Comment by TurboTurtle** [comment]
> > > @lucas-benedito can we add also `podman images` and `podman ps`? Would be helpful I believe to see that. Maybe `podman info` as well.
> > 
> > 
> > These are already captured by the `podman` plugin.
> 
> not for awx user where all processes and images are running for AAP

You'd need to use the `containers_common` plugin for that, e.g. `sos report -k containers_common.rootlessusers=awx`, or we could hook that up to a preset for AAP as well.

For reference, everything in sos is collected by root. All collections from this plugin would be as root, and plugins _generally_ don't have the ability to change their execution context. Even the above plugin uses `machinectl` to run in the context of the specified user(s), but it is still being executed as root.

**Comment by lucas-benedito** [comment]
> Curious on this, and I could be wrong here.
> 
> Isn't Ansible Tower/AWX something that can be used anywhere else. It seems like these are the main things we are collecting here. i.e. AWX specifically could be potentially be used on other distros, so we could make the name of the plugin more generic so that it can be used in other distros if needed in the future. I know ubuntu doesn't have packages yet, but could happen

Based on the information we have, there are no plans for making the RPMs available for platforms other than RHEL.

File: sos/report/plugins/aap_controller.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            "/var/lib/awx/venv/ansible/bin/pip freeze"
+            "/var/lib/awx/venv/ansible/bin/pip freeze -l"
+            "tree -d /var/lib/awx"
+            "ls -ll /var/lib/awx"
+            "ls -ll /var/lib/awx/venv"
+            "ls -ll /etc/tower",
+            "umask -p",
+        ])
+
+        def postproc(self):
```
[the remark this answers is not in the document]
Comment (lucas-benedito) [reply sos/report/plugins/aap_controller.py]: Done.

**Review by pmoravec — APPROVED** [review]

**Comment by lucas-benedito** [comment]
> > > > @lucas-benedito can we add also `podman images` and `podman ps`? Would be helpful I believe to see that. Maybe `podman info` as well.
> > > 
> > > 
> > > These are already captured by the `podman` plugin.
> > 
> > 
> > not for awx user where all processes and images are running for AAP
> 
> You'd need to use the `containers_common` plugin for that, e.g. `sos report -k containers_common.rootlessusers=awx`, or we could hook that up to a preset for AAP as well.
> 
> For reference, everything in sos is collected by root. All collections from this plugin would be as root, and plugins _generally_ don't have the ability to change their execution context. Even the above plugin uses `machinectl` to run in the context of the specified user(s), but it is still being executed as root.

@TurboTurtle what would be the best approach to hook it up to AAP to reduce the need for end-user interaction?

**Comment by TurboTurtle** [comment]
> @TurboTurtle what would be the best approach to hook it up to AAP to reduce the need for end-user interaction?

Just the collection of the `containers_common` plugin for the `awx` user? A preset would be best. If you're looking for non-root globally, that's not something we support - non-root report generation was dropped with sos-4.0. `sos collect` (controlled parallel collections over multiple nodes at once) can execute as non-root to facilitate the collections, but it still needs to become root on the nodes it is generating reports on.

A preset would allow us to specify certain conditions that make us confident we are on an AAP node, and then tack on a pre-defined set of options for any execution of `sos report`. Usually this is by the presence of certain packages, so in this case if we highlight one or more AAP-specific packages we can create a preset that effective says "when these packages are present on the host, add `-k containers_common.rootlessusers=awx`  (and anything else you'd want the be standard use) to any options the user passes"

**Review by TurboTurtle — APPROVED** [review]

**Pull request merged**